editor/src/messages/tool/tool_messages/path_tool.rs

```diff
@@ -264,14 +440,41 @@ impl Fsm for PathToolFsmState {
 						}
 					}
 
-					// Move the selected points by the mouse position
 					let snapped_position = tool_data.snap_manager.snap_position(responses, document, input.mouse.position);
+					let axis_aligned_position = axis_align_drag(shift_pressed, snapped_position, tool_data.drag_start_pos);
 					tool_data
 						.shape_editor
-						.move_selected_points(snapped_position - tool_data.previous_mouse_position, shift_pressed, responses);
-					tool_data.previous_mouse_position = snapped_position;
+						.move_selected_points(axis_aligned_position - tool_data.previous_mouse_position, shift_pressed, responses);
+					tool_data.previous_mouse_position = axis_aligned_position;
 					PathToolFsmState::Dragging
 				}
+
+				(_, PathToolMessage::BeginGrab) => {
+					tool_data.previous_mouse_position = input.mouse.position;
+					tool_data.drag_start_pos = input.mouse.position;
+
+					PathToolFsmState::Dragging
+				}
+				(_, PathToolMessage::BeginRotate) => {
+					// TODO: need start pos of mouse to calculate angle
+					tool_data.grs_mouse_start = input.mouse.position;
+
+					let path = tool_data.shape_editor.selected_layers_ref();
+					let viewspace = &mut document.document_legacy.generate_transform_relative_to_viewport(path[0]).ok().unwrap();
+
+					PathToolFsmState::Rotating
+				}
+				(_, PathToolMessage::BeginScale) => {
```

Comment:
You need to reset `tool_data.factor = 1` when you begin a scale to avoid jumping.